### DISCUSSION TITLE
How to Use FIFO(IPC meachanism) on QNX

### DISCUSSION THREAD

**Discussion by Vishalwagh05** [COVESA/dlt-daemon#560] [Q&A] [answered]
Hello Team,

I want to install dlt-daemon with FIFO on QNX, so how to do it? which flags has to set in cmakelists while cross compiling dlt on linux? what permissions are needed for FIFO file because currently after running dlt-daemon binary on QNX then I faced this issue:

DLT~1351740~WARNING  ~FIFO user /var/dlt cannot be opened (Operation not permitted)!
DLT~1351740~ERROR    ~Unable to initialize fifo.
DLT~1351740~CRITICAL ~Initialization of local connections failed!

How to resolve this?

**Comment by minminlittleshrimp** [comment]
Hello @Vishalwagh05 

I afraid this issue from yours is actually out of our scope.
There are some reasons:
+ QNX is now provided close-source.
+ Depending on your project, target, chipset, dir that allows you to open FIFO socket there!?
+ Specific permission provided to you, obviously there is a message (not permitted)

Regards

**Comment by Vishalwagh05** [comment]
Okay thank you @minminlittleshrimp

**Comment by Vishalwagh05** [comment]
@minminlittleshrimp Actually issue is with security policies of QNX so after changing those policies now I am able to run dlt-daemon on QNX...

but now when I just running simple DLT_Daemon code on QNX then I faced this issue:

[ 1196.742587]~DLT~1294395~NOTICE   ~[dlt_free] data is still readable... [14] bytes read

so @minminlittleshrimp you have any idea about it then please suggest me....

**Comment by minminlittleshrimp** [comment]
Hello @Vishalwagh05 
I believe this is not related to FIFO QNX.
It points out there are readable bytes when trying to free using UNIX_SOCKET:
https://github.com/COVESA/dlt-daemon/blob/f74e548ba3b7cf989d4b80bcb918f31840d980ae/src/lib/dlt_user.c#L1095

**Comment by Vishalwagh05** [comment]
Hello @minminlittleshrimp, Yes it is not related to FIFO QNX.
I am just running simple dlt-daemon code on QNX, where I faced this issue.
Any solution for that?

**Comment by minminlittleshrimp** [comment]
Hello,
it is not clear to me about dlt code, is it an App you build that using libdlt?

**Comment by Vishalwagh05** [comment]
Hello @minminlittleshrimp , Yes I build this with libdlt.

**Comment by minminlittleshrimp** [comment]
I do not think this is an issue, since the setting up and testing on our side.
I would suggest a briefly reading through developer's doc: https://github.com/COVESA/dlt-daemon/blob/master/doc/dlt_for_developers.md
There are things to avoid when you are trying to perform a logging on QNX with DLT.
Regards

**Reply by minminlittleshrimp** [reply]
I am not so sure about FIFO QNX since I only tested with UNIX_SOCKET and it fines.
Maybe I need to test with FIFO to reproduce this.

**Comment by minminlittleshrimp** [comment]
Transfer to discussion, please notice.
Thank you!

**Comment by Vishalwagh05** [comment]
Hello @minminlittleshrimp ,Because of FIFO issue, I am configuring dlt-daemon with UNIX_SOCKET only.
With this I faced this issue.
The example is mentioned in the link that example only I am running[ https://github.com/COVESA/dlt-daemon/blob/master/doc/dlt_for_developers.md](url) but still same issue.
- How I run this example:
firstly I run "dlt-daemon -d" with this dlt-daemon running in the background and then I run my cross compiled sample code binary.

**Reply by minminlittleshrimp** [reply]
Okay then I will try with FIFO, it is better to discuss based on these docs:

https://www.qnx.com/developers/docs/6.3.2/neutrino/lib_ref/m/mkfifo.html

And: https://www.qnx.com/developers/docs/6.3.2/neutrino/lib_ref/s/stat.html#AccessPermissions

I think it would help to solve.

**Comment by Vishalwagh05** [comment]
Hello @minminlittleshrimp , But I didn't want to run with FIFO because with UNIX_SOCKET its working. But as i mentioned I face issue when running example after running "./dlt-daemon -d" binary.
issue= "dlt_vlog(LOG_NOTICE, "[%s] data is still readable... [%zd] bytes read\n",

**Comment by minminlittleshrimp** [comment, marked as answer]
Hello @Vishalwagh05 
After a deep studying of libdlt and dlt-daemon IPC, I could point our my insight here. For QNX, due to the situation of a complex pid tree with many parents and their childs forked from, pids are not unique anymore. dlt FIFO named pipes will be created for each user based on their pid, this would not be allowed since dlt needs then do a lot of reregistration for the same Appid/Ctxid causing locking and decrease performance. Hence, dlt currently still does not support FIFO on the QNX neutrino microkernel. So, Unix socket is better here and is a recommendation for logging with dlt on this RTOS.
Regards
